### ISSUE TITLE
[red-knot] Non-deferred self-references in annotations

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ruff#16341]
### Description

We currently do not emit a diagnostic for something like the following, where a class refers to itself in a type annotation (without `from __future__ import annotations`, not in a stub file, and without stringifying it as `"C"`):
```py
class C:
    def f(self) -> C:
        ...
```
This leads to a `NameError` at runtime, so we should ideally detect it.

**Comment by carljm** [comment]
It looks like there are two separate bugs involved here; see https://github.com/astral-sh/ruff/pull/16861#discussion_r2006569021
